README.md
# xss-filter
> xss-filter is a XSS (Cross-Site Script) Filter for Node.js & the browser, provides friendly, reliable XSS filter API for you.

[![Build Status](https://travis-ci.org/superRaytin/xss-filter.svg?branch=master)](https://travis-ci.org/superRaytin/xss-filter)
[![NPM version][npm-image]][npm-url]
[![Downloads][downloads-image]][npm-url]
[![Bower version][bower-image]][bower-url]

[![xssfilter](https://nodei.co/npm/xssfilter.png)](https://npmjs.org/package/xssfilter)

[npm-url]: https://npmjs.org/package/xssfilter
[downloads-image]: http://img.shields.io/npm/dm/xssfilter.svg
[npm-image]: http://img.shields.io/npm/v/xssfilter.svg
[bower-url]:http://badge.fury.io/bo/xssFilter
[bower-image]: https://badge.fury.io/bo/xssFilter.svg

[中文](https://github.com/superRaytin/xssFilter/blob/master/README-CN.md)

[API Documentation](#manifest)

Test HTML:

```html
<div class ="like" ondblclick= "ondblclick(); return false;" onmousedown="mousedown()">
	<div class="title" title="I am a title!" value = "big">title</div>
	<div class="desc" onsubmit="load()">desc</div>
	<div>just a div</div>
	<style type="text">
		.red{color: #f00}
	</style>
	<script>alert(88)</script>
</div>
<script>alert(99)</script>
```

Result in：

```html
<div class="like">
	<div class="title" title="I am a title!" value="big">title</div>
	<div class="desc">desc</div>
	<div>just a div</div>
</div>
```

# Installation / Download

`npm install xssfilter` or `bower install xssFilter` or just download [xssFilter.js](dist/xssFilter.js) from the git repo.

# Usage

### Node.js

```js
var xssFilter = require('xssfilter');
var xssfilter = new xssFilter();

var output = xssfilter.filter('<div class="like" ondblclick="takeme()" onmousedown="mousedown()">something...</div>');
// output: <div class="like">something...</div>
```

### Browser

```js
<script src="./dist/xssFilter.js"></script>
<script>
    var xssfilter = new xssFilter();
    var output = xssfilter.filter('<div class="like" ondblclick="takeme()" onmousedown="mousedown()">something...</div>');
    // output: <div class="like">something...</div>
</script>
```

#### Use with [require.js](http://requirejs.org/)

```js
<script src="require.js"></script>
<script>
    define(function() {
        var xssFilter = require('./dist/xssFilter.js');
        var xssfilter = new xssFilter();

        var output = xssfilter.filter('<div class="like" ondblclick="takeme()" onmousedown="mousedown()">something...</div>');
        // output: <div class="like">something...</div>
    });
</script>
```

#### Use with [sea.js](http://seajs.org/)

```js
<script src="sea.js"></script>
<script>
    seajs.use('./dist/xssFilter.js', function(xssFilter){
        var xssfilter = new xssFilter();

        // "<" to &lt; ">" to &gt;
        xssfilter.options('escape', true);

        var output = xssfilter.filter('<div class="like" ondblclick="takeme()" onmousedown="mousedown()">something...</div>');
        // output: &lt;div class="like"&gt;something...&lt;/div&gt;
    })
</script>
```

# Manifest

### matchStyleTag

whether match `style` tag, default is `true`. Set to `false` to prevent remove the matched `style` tags.

### matchScriptTag

whether match `script` tag, default is `true`. Set to `false` to prevent remove the matched `script` tags.

### removeMatchedTag

whether remove matched tag, default is `true`. Set to `false` to using escape instead of remove.

`removeMatchedTag` should be used with `matchStyleTag` and `matchScriptTag`, for example:

```js
var xssfilter = new xssFilter({
    removeMatchedTag: false
});
```

```html
<div class ="like" onmousedown="mousedown()">
	<style type="text">
		.red{color: #f00}
	</style>
	something...
</div>
<script>alert(88)</script>
```

Result in:

```html
<div class="like">
	&lt;style type="text"&gt;
		.red{color: #f00}
	&lt;/style&gt;
	something...
</div>
&lt;script&gt;alert(88)&lt;/script&gt;
```

### blackListAttrs

attributes blacklist, attributes in this list will be cleared.

initial blacklist of attributes：

```js
{
    onclick: true,
    ondblclick: true,
    onchange: true,
    onblur: true,
    onfocus: true,
    onkeydown: true,
    onkeypress: true,
    onkeyup: true,
    onmousedown: true,
    onmousemove: true,
    onmouseover: true,
    onmouseout: true,
    onmouseup: true,
    onselect: true,
    onsubmit: true,
    onreset: true,
    onload: true,
    onabort: true,
    onerror: true
}
```

### escape

escape tags of whole html string, `"<" to "&lt;", ">" to "&gt;"`, default no.


# Initialization
The configuration options can be specified by passing an `options` parameter in the initialization. `options` is optional, provided to override the default configuration.

```js
var xssfilter = new xssFilter(options);
```

# Instance methods

### filter
Filtering target string, accepts only one parameter.

### options

Use this method to modify the configuration options after initialization.

```js
var xssfilter = new xssFilter();

xssfilter.options({
    escape: true,
    matchStyleTag: false
});

var output = xssfilter.filter('some html...');
```

You can also configure single option:

```js
var xssfilter = new xssFilter();
xssfilter.options('escape', true);
var output = xssfilter.filter('some html...');
```

when set secondary attributes like `blackListAttrs`, the second argument must be an object `{}`:

```js
var xssfilter = new xssFilter();

xssfilter.options('blackListAttrs', {
    onsubmit: false
});

var output = xssfilter.filter('<div class="like" ondblclick="ondblclick();" onsubmit="dosomething()">something...</div>');
// output: <div class="like" onsubmit="dosomething()">something...</div>
```

# Testing

```
npm test
```

# Other xss filter view

- [https://github.com/leizongmin/js-xss](https://github.com/leizongmin/js-xss)
- [https://github.com/yahoo/xss-filters](https://github.com/yahoo/xss-filters)

# License
MIT, see the [LICENSE](https://github.com/superRaytin/xssFilter/blob/master/LICENSE) file for detail.


CHANGELOG.md
## 0.5.3

- Fix: attribute in black list should be case insensitive #1

## 0.5.2

- Support using in browser

## 0.5.1

- Add test cases
- Add `removeMatchTag`, use to remove matched tag, if set to false, escape instead.
- Optimize clear tag spaces method

## 0.5.0

- Add browserify support
- Reconstruction & optimization

test/utils.js

var should = require('should');
var utils = require('../lib/utils');

describe('utils', function() {
  it('should str_trim() work', function() {
    utils.str_trim(' hello').should.equal('hello');
    utils.str_trim('hello').should.equal('hello');
    utils.str_trim(' hello ').should.equal('hello');
    utils.str_trim('      hello      ').should.equal('hello');
  });

  it('should arr_compact() work', function() {
    utils.arr_compact(['', 'a', 'b', '', 'c', '']).should.eql(['a', 'b', 'c']);
    utils.arr_compact(['', 'a', 'b', 'c', '']).should.eql(['a', 'b', 'c']);
  });

  it('should isObject() work', function() {
    utils.isObject({a: 1}).should.equal(true);
    utils.isObject(['a']).should.equal(true);
  });

  it('should extend() work', function() {
    var obj = {a: 1};
    utils.extend(obj, {b: 2});
    obj.should.eql({a: 1, b: 2});

    obj = {a: 1, b: 6};
    utils.extend(obj, {a: 3});
    obj.should.eql({a: 3, b: 6});
  });
});

test/xssfilter.js

var should = require('should');
var xssFilter = require('../lib/index');

var htmlHead = '<div class="like"  ondblclick= "ondblclick(); return false;"  onmousedown="mousedown()">';
var text = 'something...';
var htmlFoot = '</div>';
var defaultScript = '<script>alert(88)</script>';
var defaultMultiLineScript = '<script>\n' +
    'alert(88)\n' +
    '</script>';
var defaultStyle = '<style type="text">.red{color: #f00}</style>';
var defaultStyleEscaped = '&lt;style type="text"&gt;.red{color: #f00}&lt;/style&gt;';
var defaultMultiLineStyle = '<style type="text">\n' +
    '.red{color: #f00}\n' +
    '</style>';
var defaultMultiLineStyleEscaped = '&lt;style type="text"&gt;\n' +
    '.red{color: #f00}\n' +
    '&lt;/style&gt;';
var div1 = '<div class="title" title="I am a title!" value = "big">title</div>';
var div2 = '<div class="desc" onsubmit="load()">desc</div>';


var defaultHTML = htmlHead + text + htmlFoot;

describe('filter', function() {
  it('should filter() work', function() {
    var xssfilter = new xssFilter();
    xssfilter.filter(defaultHTML).should.equal('<div class="like">something...</div>');
    xssfilter.options('blackListAttrs', {
      onmousedown: false
    });
    xssfilter.filter('<div class="like"     ondblclick= "ondblclick(); return false;"       onmousedown  ="mousedown()"    >something...</div>').should.equal('<div class="like" onmousedown="mousedown()">something...</div>');
    xssfilter.filter('<div class="like"     ondblclick= "ondblclick(); return false;"       onmousedown  ="mousedown()" data-title="onmousedown=mousedown()"    >something...</div>').should.equal('<div class="like" onmousedown="mousedown()" data-title="onmousedown=mousedown()">something...</div>');
  });

  it('should options() work', function() {
    var xssfilter = new xssFilter();
    xssfilter.options({
      escape: true
    });
    xssfilter.config.escape.should.equal(true);

    xssfilter.options({
      escape: false
    });
    xssfilter.config.escape.should.equal(false);
  });

  it('should escape work', function() {
    var xssfilter = new xssFilter();
    xssfilter.options('escape', true);
    xssfilter.filter(defaultHTML).should.equal('&lt;div class="like"&gt;something...&lt;/div&gt;');

    xssfilter.options('escape', false);
    xssfilter.filter(defaultHTML).should.equal('<div class="like">something...</div>');
  });

  it('should blackListAttrs work', function() {
    var xssfilter = new xssFilter();
    xssfilter.options('blackListAttrs', {
      ondblclick: false
    });
    xssfilter.filter(defaultHTML).should.equal('<div class="like" ondblclick="ondblclick(); return false;">something...</div>');

    xssfilter.options('blackListAttrs', {
      onmousedown: false
    });
    xssfilter.filter(defaultHTML).should.equal('<div class="like" ondblclick="ondblclick(); return false;" onmousedown="mousedown()">something...</div>');

    xssfilter.options('blackListAttrs', {
      ondblclick: true,
      onmousedown: false
    });
    xssfilter.filter(defaultHTML).should.equal('<div class="like" onmousedown="mousedown()">something...</div>');
  });

  it('should cleanTag work', function() {
    var xssfilter = new xssFilter();
    xssfilter.filter(defaultHTML).should.equal('<div class="like">something...</div>');
    xssfilter.filter('<div    class="like"     ondblclick= "ondblclick();      return false;"     onmousedown    =           "mousedown();     alert(8888)" aaa bbb   >something...</div>').should.equal('<div class="like" aaa bbb>something...</div>');
  });

  it('should removeMatchedTag work', function() {
    var xssfilter = new xssFilter();
    var html = htmlHead + text + defaultStyle + htmlFoot;
    xssfilter.filter(html).should.equal('<div class="like">something...</div>');

    xssfilter.options('removeMatchedTag', false);
    xssfilter.filter(html).should.equal('<div class="like">something...'+ defaultStyleEscaped +'</div>');

    var html2 = htmlHead + text + defaultMultiLineStyle + htmlFoot;
    xssfilter.options('removeMatchedTag', false);
    xssfilter.filter(html2).should.equal('<div class="like">something...'+ defaultMultiLineStyleEscaped +'</div>');
  });

  it('should matchStyleTag work', function() {
    var xssfilter = new xssFilter();

    xssfilter.options('matchStyleTag', false);
    var html = htmlHead + text + defaultStyle + htmlFoot;
    xssfilter.filter(html).should.equal('<div class="like">something...'+ defaultStyle +'</div>');

    xssfilter.options('matchStyleTag', false);
    html = htmlHead + text + defaultMultiLineStyle + htmlFoot;
    xssfilter.filter(html).should.equal('<div class="like">something...'+ defaultMultiLineStyle +'</div>');

    xssfilter.options('matchStyleTag', true);
    html = htmlHead + text + defaultStyle + htmlFoot;
    xssfilter.filter(html).should.equal('<div class="like">something...</div>');

    xssfilter.options('matchStyleTag', true);
    html = htmlHead + text + defaultMultiLineStyle + htmlFoot;
    xssfilter.filter(html).should.equal('<div class="like">something...</div>');
  });

  it('should matchScriptTag work', function() {
    var xssfilter = new xssFilter();

    xssfilter.options('matchScriptTag', false);
    var html = htmlHead + text + defaultScript + htmlFoot;
    xssfilter.filter(html).should.equal('<div class="like">something...'+ defaultScript +'</div>');

    xssfilter.options('matchScriptTag', false);
    html = htmlHead + text + defaultMultiLineScript + htmlFoot;
    xssfilter.filter(html).should.equal('<div class="like">something...'+ defaultMultiLineScript +'</div>');

    xssfilter.options('matchScriptTag', true);
    html = htmlHead + text + defaultScript + htmlFoot;
    xssfilter.filter(html).should.equal('<div class="like">something...</div>');

    xssfilter.options('matchScriptTag', true);
    html = htmlHead + text + defaultMultiLineScript + htmlFoot;
    xssfilter.filter(html).should.equal('<div class="like">something...</div>');
  });

  it('should work when attribute value was double quoted or single quoted or unquoted', function() {
    var xssfilter = new xssFilter();
    xssfilter.filter('<div class="like" onmousedown="mousedown()">something...</div>').should.equal('<div class="like">something...</div>');
    xssfilter.filter("<div class='like' onmousedown='mousedown()'>something...</div>").should.equal("<div class='like'>something...</div>");
    xssfilter.filter('<div class=likebbbbbbbbbbbbbb onmousedown=mousedown()>something...</div>').should.equal('<div class=likebbbbbbbbbbbbbb>something...</div>');
  });

  it('should work when there are spaces between attribute name and "=", like class ="like"', function() {
    var xssfilter = new xssFilter();
    xssfilter.filter('<div class ="like" ondblclick= "ondblclick(); return false;" onmousedown="mousedown()">something...</div>').should.equal('<div class="like">something...</div>');
    xssfilter.filter('<div class = "like" ondblclick= "ondblclick(); return false;" onmousedown="mousedown()">something...</div>').should.equal('<div class="like">something...</div>');
    xssfilter.filter('<div class =  "like" ondblclick= "ondblclick();    return false;" onmousedown="mousedown()">something...</div>').should.equal('<div class="like">something...</div>');
  });

  it('should work when attribute is uppercase', function() {
    var xssfilter = new xssFilter();
    xssfilter.filter('<div class ="like" onMouseOver= "dosomething(); return false;" onmousedown="mousedown()">something...</div>').should.equal('<div class="like">something...</div>');
  });
});

lib/index.js

var utils = require('./utils');

// 匹配 style 标签的正则表达式
var REGEXP_TAG_STYLE = /<style[^>]*>[^<]*<\/style>/img;

// 匹配 script 标签的正则表达式
var REGEXP_TAG_SCRIPT = /<script[^>]*>[^<]*<\/script>/img;

// 匹配标签中的属性 的正则表达式
var REGEXP_ATTR = /([\w-]+)\s*=\s*("([^"]*)"|'([^']*)'|([^ >]*))/img;

// 匹配带有属性并且属性名称前面有空格的标签 的正则表达式
var REGEXP_ATTR_SPACE_PRE = /\s*([\w-]+)\s*=\s*("([^"]*)"|'([^']*)'|([^ >]*))/img;

// 匹配带有属性并且属性值后面有空格的标签 的正则表达式
var REGEXP_ATTR_SPACE_SUF = /([\w-]+)\s*=\s*("([^"]*)"|'([^']*)'|([^ >]*))(\s*)/img;

// 匹配带有属性的标签 的正则表达式
var REGEXP_ATTR_WITH_TAG = /<[a-zA-Z]+[a-zA-Z0-9]*((\s+([\w-]+)\s*=\s*("([^"]*)"|'([^']*)'|([^ >]*)))+).*>/img;

/**
 * 构造函数
 *
 * @param {Object} options, 配置对象
 * @return {String}
 * */
function XSSFilter(options) {
    // 初始化配置
    this.config = {
        // 是否匹配 style 标签
        matchStyleTag: true,

        // 是否匹配 script 标签
        matchScriptTag: true,

        // 是否要删除匹配到的标签，如果设置为 false，则对标签进行转义
        removeMatchedTag: true,

        // 是否对整个字符串进行转义，默认不转义
        escape: false,

        // 标签属性黑名单列表，在这个列表中的属性将被清除
        blackListAttrs: {
            onclick: true,
            ondblclick: true,
            onchange: true,
            onblur: true,
            onfocus: true,
            onkeydown: true,
            onkeypress: true,
            onkeyup: true,
            onmousedown: true,
            onmousemove: true,
            onmouseover: true,
            onmouseout: true,
            onmouseup: true,
            onselect: true,
            onsubmit: true,
            onreset: true,
            onload: true,
            onabort: true,
            onerror: true
        }
    };

    // 实例化时扩展初始化配置
    if (utils.isObject(options)) {
        utils.extend(this.config, options);
    }
}


/**
 * 修改过滤参数的方法
 *
 * @param {String} name
 * @param {String} obj
 *
 * 支持 3 种调用方式：
 * filterConstant.options('escape', true);
 *
 * filterConstant.options({
 *  escape: true
 * });
 *
 * filterConstant.options('blackListAttrs', {
 *  onclick: false
 * })
 *
 * @return {String}
 * */
XSSFilter.prototype.options = function(name, obj) {
    var config = this.config;

    if (arguments.length) {
        if (typeof name === 'string') {
            if (typeof config[name] === 'undefined') {
                throw new Error(name + ' is not a valid configuration name.');
            }

            if (typeof obj === 'undefined') {
                throw new Error('Please enter a value corresponding to the ' + name);
            }

            if (utils.isObject(obj)) {
                utils.extend(config[name], obj);
            } else {
                config[name] = obj;
            }
        } else if (utils.isObject(name)) {
            obj = name;
            utils.extend(config, obj);
        }
    }
};


/**
 * 过滤 html 的方法
 *
 * @param {String} html
 * @return {String}
 * */
XSSFilter.prototype.filter = function(html) {
    if (html == '') return html;

    var result = html;
    var config = this.config;

    if (config.matchStyleTag) {
        result = filterStyleTag(result, config);
    }

    if (config.matchScriptTag) {
        result = filterScriptTag(result, config);
    }

    result = filterAttribute(result, config);

    // clear tag spaces
    result = clearTagSpaces(result);

    if (config.escape) {
        result = escapeTags(result);
    }

    return result;
};


/**
 * 过滤在黑名单中的标签属性
 *
 * @param {String} html, 原始的 html 字符串
 * @param {Object} config, 配置对象
 * @return {String}
 * */
function filterAttribute(html, config) {
    var result = html;
    var tempHTML = html;

    (function() {
        //
        var attrMatches = REGEXP_ATTR_WITH_TAG.exec(tempHTML);
        REGEXP_ATTR_WITH_TAG.lastIndex = 0;

        if (attrMatches) {
            var labelHasAttr = attrMatches[1];
            var attrArray = labelHasAttr.match(REGEXP_ATTR);

            tempHTML = tempHTML.replace(labelHasAttr, '');

            utils.each(attrArray, function(item) {
                var attrName = utils.str_trim(item.substr(0, item.indexOf('='))).toLowerCase();

                if (config.blackListAttrs[attrName]) {
                    result = result.replace(item, '');
                }
            });

            arguments.callee();
        }
    })();

    return result;
}


/**
 * 过滤 style 标签
 *
 * @param {String} html
 * @param {Object} config, 配置对象
 * @return {String}
 * */
 function filterStyleTag(html, config) {
    var result;

    if (config.removeMatchedTag) {
        result = html.replace(REGEXP_TAG_STYLE, '');
    } else {
        result = html.replace(REGEXP_TAG_STYLE, function(body) {
            return escapeTags(body);
        });
    }

    return result;
}


/**
 * 过滤 script 标签
 *
 * @param {String} html
 * @param {Object} config, 配置对象
 * @return {String}
 * */
function filterScriptTag(html, config) {
    var result;

    if (config.removeMatchedTag) {
        result = html.replace(REGEXP_TAG_SCRIPT, '');
    } else {
        result = html.replace(REGEXP_TAG_SCRIPT, function(body) {
            return escapeTags(body);
        });
    }

    return result;
}


/**
 * 清理标签中多余的空格
 *
 * @param {String} html
 * @return {String}
 * */
function clearTagSpaces(html) {

    // 主要移除三种空格
    // 1. 属性名称前面多余的空格
    // 2. 属性值后面多余的空格
    // 3. 等号两边的空格
    // 示例：[   name  =  "value"  ] -> [ name="value" ]
    var result = html.replace(REGEXP_ATTR_SPACE_PRE, function(dirtyAttr, attrName, attrValue) {
        return ' ' + attrName + '=' + attrValue;
    }).replace(REGEXP_ATTR_SPACE_SUF, function(dirtyAttr, attrName, attrValue, a, b, c, sufSpace) {
        var cleanAttr = attrName + '=' + attrValue;

        // 如果结尾没有空格（sufSpace == 0），则不补加单个空字符串
        if (sufSpace.length > 0) {
            cleanAttr += ' ';
        }

        return cleanAttr;
    });

    // 移除标签结束符前面的空格
    // 示例：[<div   >] -> [<div>]
    result = result.replace(/\t+\n/g, '').replace(/\s*>/mg, function(a) {
        return a.replace(/\s+/, '');
    });

    return result;
}


/**
 * 对标签进行转义
 *
 * @param {String} html
 * @return {String}
 * */
function escapeTags(html) {
    return html.replace(/</g, '&lt;').replace(/>/g, '&gt;');
}

// 在 AMD 模块环境中使用
if (typeof define === 'function' && define.amd) {
    define(function(){
        return XSSFilter;
    });
}
// 在浏览器中使用
else if (typeof window !== 'undefined') {
    window.xssFilter = XSSFilter;
}

module.exports = XSSFilter;


lib/utils.js
/*
 * utils
 */

var utils = {
    each: function(stack, handler) {
        var len = stack.length;
        if (len) {
            for(var i = 0; i < len; i++) {
                if (handler.call(stack[i], stack[i], i) === false) break;
            }
        }
        else if (typeof len === 'undefined') {
            for(var name in stack) {
                if (handler.call(stack[name], stack[name], name) === false) break;
            }
        }
    },

    str_trim: function(string) {
        return string.replace(/^\s+/g, '').replace(/\s+$/g, '');
    },

    arr_compact: function(array) {
        var result = [];
        utils.each(array, function(item) {
            if (utils.str_trim(item) != '') {
                result.push(item);
            }
        });

        return result;
    },

    isObject: function(obj) {
        return obj === Object(obj);
    },

    extend: function(target, obj) {
        utils.each(obj, function(value, key) {
            target[key] = value;
        });
    }
};

module.exports = utils;



xssFilter.js
/*
 * xssFilter
 * https://github.com/superRaytin/xssFilter
 *
 * Copyright 2014, SuperRaytin
 * Released under the MIT license.
 */

(function(global){
    // RegExp
    var REGEXP_LABEL_STYLE = /<style[^>]*>[^<]*<\/style>/img;
    var REGEXP_LABEL_SCRIPT = /<script[^>]*>[^<]*<\/script>/img;
    var REGEXP_ATTR = /([\w-]+)\s*=\s*("([^"]*)"|'([^']*)'|([^ ]*))/img;
    var REGEXP_ATTR_LABEL = /<[a-zA-Z]+((\s+([\w-]+)\s*=\s*("([^"]*)"|'([^']*)'|([^ ]*)))+)>/img;

    // Tools
    var utils = {
        each: function(stack, handler){
            var len = stack.length;
            if(len){
                for(var i = 0; i < len; i++){
                    if(handler.call(stack[i], stack[i], i) === false) break;
                }
            }
            else if(typeof len === 'undefined'){
                for(var name in stack){
                    if(handler.call(stack[name], stack[name], name) === false) break;
                }
            }
        },
        str_trim: function(string){
            return string.replace(/^\s+/g, '').replace(/\s+$/g, '');
        },
        arr_compact: function(array){
            var result = [];
            utils.each(array, function(item){
                if(utils.str_trim(item) != ''){
                    result.push(item);
                }
            });

            return result;
        },
        isObject: function(obj){
            return obj === Object(obj);
        },
        extend: function(target, obj){
            utils.each(obj, function(value, key){
                target[key] = value;
            });
        }
    };

    // constructor
    function XSSFilter(options){
        if(utils.isObject(options)){
            utils.extend(this.config, options);
        }
    }

    // Initial Configuration
    XSSFilter.prototype.config = {
        // filter style label
        label_style: true,

        // filter script label
        label_script: true,

        // Beautify Tags
        beautifyTags: true,

        // Escape
        escape: false,

        blackList_attrs: {
            onclick: true,
            ondblclick: true,
            onchange: true,
            onblur: true,
            onfocus: true,
            onkeydown: true,
            onkeypress: true,
            onkeyup: true,
            onmousedown: true,
            onmousemove: true,
            onmouseover: true,
            onmouseout: true,
            onmouseup: true,
            onselect: true,
            onsubmit: true,
            onreset: true,
            onload: true,
            onabort: true
        }
    };

    /*
     * Filter Options Configuration
     * */
    XSSFilter.prototype.options = function(name, obj){
        var args = arguments;
        var config = this.config;

        if(args.length){
            if(typeof name === 'string'){
                if(typeof config[name] === 'undefined'){
                    throw new Error(name + ' is not a valid configuration name.');
                    return;
                }

                if(typeof obj === 'undefined'){
                    throw new Error('Please enter a value corresponding to the ' + name);
                    return;
                }

                if(utils.isObject(obj)){
                    utils.extend(config[name], obj);
                }
                else{
                    config[name] = obj;
                }
            }
            else if(utils.isObject(name)){
                obj = name;
                utils.extend(config, obj);
            }
        }
    };

    /*
     * Filter Attributes in Blacklist
     * */
    var _filterAttribute = function(html, config){
        var result = html;
        var tempHTML = html;

        (function(){
            var attrMatches = REGEXP_ATTR_LABEL.exec(tempHTML);
            REGEXP_ATTR_LABEL.lastIndex = 0;

            if(attrMatches){
                var labelHasAttr = attrMatches[1];
                var attrArray = labelHasAttr.match(REGEXP_ATTR);

                tempHTML = tempHTML.replace(labelHasAttr, '');

                utils.each(attrArray, function(item){
                    var attrName = utils.str_trim(item.substr(0, item.indexOf('=')));

                    if(config.blackList_attrs[attrName]){
                        result = result.replace(item, '');
                    }
                });

                arguments.callee();
            }
        })();

        return result;
    };

    /*
     * Filter Style Tag
     * */
    var _filterLabelStyle = function(candy){
        return candy.replace(REGEXP_LABEL_STYLE, '');
    };

    /*
     * Filter Script Tag
     * */
    var _filterLabelSript = function(candy){
        return candy.replace(REGEXP_LABEL_SCRIPT, '');
    };

    /*
    * Beautify Tags
    * */
    var _beautifyTags = function(candy){
        return candy.replace(/\t+\n/g, '').replace(/\s*>/mg, function(a){
            return a.replace(/\s+/, '');
        });
    };

    /*
     * Escape Tags
     * */
    var _escapeTags = function(candy){
        return candy.replace(/</g, '&lt;').replace(/>/g, '&gt;');
    };

    XSSFilter.prototype.filter = function(html){
        if(html == ''){
            return;
        }

        var result = html;
        var config = this.config;

        if(config.label_style){
            result = _filterLabelStyle(result);
        }

        if(config.label_script){
            result = _filterLabelSript(result);
        }

        result = _filterAttribute(result, config);

        if(config.beautifyTags){
            result = _beautifyTags(result);
        }

        if(config.escape){
            result = _escapeTags(result);
        }

        return result;
    };

    /*
    * export via AMD or CommonJS, otherwise leak a global
    * */
    if(typeof define === 'function' && (define.amd || define.cmd)){
        define(function(){
            return XSSFilter;
        });
    }
    else if(typeof module !== 'undefined' && typeof module.exports !== 'undefined'){
        module.exports = XSSFilter;
    }
    else{
        global.xssFilter = XSSFilter;
    }

})(this);


example/example.js
var xssFilter = require('../lib/index');

var testHTML =  '<div class ="like" ondblclick= "ondblclick(); return false;" onmousedown="mousedown()">\n' +
                    '\t<div class="title" title="I am a title!" value = "big">title</div>\n' +
                    '\t<div class="desc" onsubmit="load()">desc</div>\n' +
                    '\t<div>just a div</div>\n' +
                    '\t<style type="text">\n' +
                        '\t\t.red{color: #f00}\n' +
                    '\t</style>\n' +
                    '\t<script>alert(88)<\/script>\n' +
                '</div>\n' +
                '<script>alert(99)<\/script>';

var xssfilter = new xssFilter();
var candy = xssfilter.filter(testHTML);

console.log('================== testing HTML ==================')
console.log(testHTML);
console.log('\n================== result ==================')
console.log(candy);